File: Core/C#/Core/Searching/BinarySearch.cs
/*
    Binary search algorithm, works for sorted arrays in O(log n)
    because half the array is continually discarded on each pass!
*/

namespace Core.Searching;

public static class BinarySearch
{
    public static int Search(int[] arr, int value)
    {
        return Search(arr, 0, arr.Length - 1, value);
    }

    // Binary search, keep halving the array untill we find
    // that x is a middle point.
    // runtime = O(log n)
    private static int Search(int[] arr, int left, int right, int x)
    {
        if (right > left)
        {
            // We know we are searching normally.
            int middle = (left + (right - 1)) / 2;
            // If middle IS x then just return that.
            if (arr[middle] == x) return middle;

            // If element is smaller than mid, then can only be present in left array,
            // and vice versa.
            // (middle - 1 = the new right for the recursive call.)
            if (x < arr[middle]) return Search(arr, left, middle - 1, x);
            return Search(arr, middle + 1, right, x);
        }
        // If no item was returned the answer is not found.
        return -1;
    }
}

File: Core/C++/Core/Searching/binary-search.cpp
/*
    Performs a simple binary search on an array of integers.
*/

int binary_search(int arr[], int value, int length)
{
    return binary_search_inner(arr, value, 0, length - 1);
}

int binary_search_inner(int arr[], int value, int left, int right)
{
    if (left > right) return -1;
    int middle = (right - left) / 2;
    if (arr[middle] == value) return middle;
    if (arr[left] == value) return left;
    if (arr[right] == value) return right;

    if (value < middle) return binary_search_inner(arr, value, left, middle - 1);
    return binary_search_inner(arr, value, middle + 1, right);
}